src/auth_code_pkce.rs

```diff
@@ -0,0 +1,158 @@
+        let scopes = oauth
+            .scope
+            .clone()
+            .into_iter()
+            .collect::<Vec<_>>()
+            .join(" ");
+        data.insert(headers::GRANT_TYPE, headers::GRANT_AUTH_CODE);
+        data.insert(headers::REDIRECT_URI, oauth.redirect_uri.as_ref());
+        data.insert(headers::CODE, code);
+        data.insert(headers::SCOPE, scopes.as_ref());
+        data.insert(headers::STATE, oauth.state.as_ref());
+
+        let token = self.fetch_access_token(&data).await?;
+        self.token = Some(token);
+
+        self.write_token_cache()
+    }
+
+    async fn refresh_token(&mut self, refresh_token: &str) -> ClientResult<()> {
+        // TODO
+        let mut data = Form::new();
+        data.insert(headers::REFRESH_TOKEN, refresh_token);
+        data.insert(headers::GRANT_TYPE, headers::GRANT_REFRESH_TOKEN);
+
+        let mut token = self.fetch_access_token(&data).await?;
+        token.refresh_token = Some(refresh_token.to_string());
+        self.token = Some(token);
+
+        self.write_token_cache()
+    }
+}
+
+impl AuthCodePkceSpotify {
+    /// Builds a new [`AuthCodePkceSpotify`] given a pair of client credentials
+    /// and OAuth information.
+    pub fn new(creds: Credentials, oauth: OAuth) -> Self {
+        AuthCodePkceSpotify {
+            creds,
+            oauth,
+            ..Default::default()
+        }
+    }
+
+    /// Build a new [`AuthCodePkceSpotify`] from an already generated token.
+    /// Note that once the token expires this will fail to make requests, as the
+    /// client credentials aren't known.
+    pub fn from_token(token: Token) -> Self {
+        AuthCodePkceSpotify {
+            token: Some(token),
+            ..Default::default()
+        }
+    }
+
+    /// Same as [`Self::new`] but with an extra parameter to configure the
+    /// client.
+    pub fn with_config(creds: Credentials, oauth: OAuth, config: Config) -> Self {
+        AuthCodePkceSpotify {
+            creds,
+            oauth,
+            config,
+            ..Default::default()
+        }
+    }
+
+    /// Returns the URL needed to authorize the current client as the first step
+    /// in the authorization flow.
+    pub fn get_authorize_url(&self) -> ClientResult<String> {
+        // TODO
+        let mut payload: HashMap<&str, &str> = HashMap::new();
+        let oauth = self.get_oauth();
+        let scope = oauth
+            .scope
+            .clone()
+            .into_iter()
+            .collect::<Vec<_>>()
+            .join(" ");
+        payload.insert(headers::CLIENT_ID, &self.get_creds().id);
+        payload.insert(headers::RESPONSE_TYPE, headers::RESPONSE_CODE);
+        payload.insert(headers::REDIRECT_URI, &oauth.redirect_uri);
+        payload.insert(headers::SCOPE, &scope);
+        payload.insert(headers::STATE, &oauth.state);
+        // payload.insert(headers::CODE_CHALLENGE, todo!());
+        // payload.insert(headers::CODE_CHALLENGE_METHOD, "S256");
```

Comment:
According to the Spotify document, `code_challenge_method` and `code_challenge` are required when constructing the authorization URI. Why would you like to comment this code? It's still WIP?

Reply:
I've left the PKCE implementation for a future PR